changefeedccl: allow users to alter changefeed options

sherman-grewal opened the pull request:
changefeedccl: allow users to alter changefeed options
with the ALTER CHANGEFEED statement

References #75895

Currently, with the ALTER CHANGEFEED statement users
can only add or drop targets from an existing
changefeed. In this PR, we would like to extend this
functionality so that an user can edit and unset
the options of an existing changefeed as well.
The syntax of this addition is the following:

ALTER CHANGEFEED <job_id> SET \<options\> UNSET <opt_list>

Note that the <options> must follow the same syntax
that is used when creating a changefeed with options.
In particular, if you would like to set an option
that requires a value, you must write

SET opt = 'value'

On the other hand, if you would like to set an
option that requires no value, you must write

SET opt

Furthermore, this PR allows users to unset options.
This can be achieved by writing

UNSET <opt_list>

Where <opt_list> is a list of options that you
would like to unset. For example, if we would like
to unset the diff and resolved options for
changefeed 123, we would achieve this by writing

ALTER CHANGEFEED 123 UNSET diff, resolved

Release note (enterprise change): Added support to
the ALTER CHANGEFEED statement so that users can edit
and unset the options of an existing changefeed. The
syntax of this addition is the following:

ALTER CHANGEFEED <job_id> SET \<options\> UNSET <opt_list>

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/76583)
<!-- Reviewable:end -->

miretskiy reviewed (changes requested):
> Reviewed 4 of 8 files at r1, 1 of 1 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76583)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @sherman-grewal, and @shermanCRL)

___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-Mw2RM1t2ewjF9cGQB3z:-Mw2RM1t2ewjF9cGQB4-:bdymqzy) ([raw file](https://github.com/cockroachdb/cockroach/blob/156f62553e2ff5f9e855d99aacf6f4faa948ef83/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L88)):*
> ```Go
> 		targetsChanged := false
> 
> 		oldStmt, err := parser.ParseOne(job.Payload().Description)
> ```

Ooof... Arguably, I should have picked this earlier (during previous version)... But here we are.

Unfortunately creating changefeed statement node based on payload description is *unsafe* and
is broken.  Basically -- Description contains redacted string representation of the original statement,
and thus cannot be reversed (e.g. you will loose any security related information -- keys, etc). -- `changefeedJobDescription` function.

What to do:
  1. Please add a test (which I expect to fail) where you create changefeed into e.g. s3, and include some of the redactable fields there (`create changefeed ... into 's3://....?AWS_SECRET_ACCESS_KEY=my_secret'`).  Verify that
after an alter statement you still see AWS_SECRET_ACCESS_KEY=my_secret and not AWS_SECRET_ACCESS_KEY=redacted.  I expect this test to fail.  
 2. You can (and probably should) fix this issue as a separate PR on top of existing alter implementation. 
 3. The changefeed statement must be synthesized based on ChangefeedDetails fields.  Now, you don't actually have 
      to regenerate exact statement as it was typed when changefeed was created.  In particular, the thing you might have problem with is the target list.  You have target id, but the name available in ChangeFeedTarget field might no longer be available -- because the table could have been renamed.  So, you can't just assume you can put statement_names in the changefeed statement.  It would be nice if we can use tree.AliasedTableExpr in changefeed statement -- but as things are right now, that's not the case (I believe TablePattern is not the same as an alias table expr).   Perhaps the changefeed statement can be changed somehow to also allow specification of targets as table expressions? If so, you could generate table expression in the form of `[id AS name]` (this is something that select supports... you can do `select * from [30 as  namespace]` -- try running this from cockroach demo).  Other possible hacks include resolving current names for each descriptor, re-synthesizing changefeed statement using current names, and then, "renaming" them to the original statement_time_name.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 123 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-Mw2eKQz2XLINo-fLJ7Z:-Mw2eKQz2XLINo-fLJ7_:b-186njc) ([raw file](https://github.com/cockroachdb/cockroach/blob/156f62553e2ff5f9e855d99aacf6f4faa948ef83/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L123)):*
> ```Go
> 		}
> 
> 		if opts.Options != nil {
> ```

So, now that we've fixed how  the statement is re-generated (above), I have to wonder... Would we be better off if we just take changefeed statement, and append all options there?

As far as I know, appending the same option multiple times overrides the previous occurrences.

 If you refactor changefeed statement plan hook so that  you have a separate function to compute the job record basically, then you could simply call into that function, and have it perform all of the validation, and return you the "new" job record which you will handle here.
The only thing is that function would probably need to take in a boolean argument indicating if it should update telemetry data  (yes, from changefeed, and now from here).
___
*[pkg/sql/parser/testdata/alter_changefeed, line 99 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-Mw2M7Jw9q9xIpq7J_OX:-Mw2M7Jw9q9xIpq7J_OY:b-j42joh) ([raw file](https://github.com/cockroachdb/cockroach/blob/156f62553e2ff5f9e855d99aacf6f4faa948ef83/pkg/sql/parser/testdata/alter_changefeed#L99)):*
> ```
> ALTER CHANGEFEED (123) ADD (foo)  SET bar = ('baz'), qux = ('quux')  DROP (corge) -- fully parenthesized
> ALTER CHANGEFEED _ ADD foo  SET bar = '_', qux = '_'  DROP corge -- literals removed
> ALTER CHANGEFEED 123 ADD _  SET _ = 'baz', _ = 'quux'  DROP _ -- identifiers removed
> ```

:heart: these tests. 
___
*[pkg/ccl/changefeedccl/show_changefeed_jobs_test.go, line 450 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-Mw2NP6C84ccmR2LUJSS:-Mw2NP6C84ccmR2LUJST:b-ntxvnf) ([raw file](https://github.com/cockroachdb/cockroach/blob/156f62553e2ff5f9e855d99aacf6f4faa948ef83/pkg/ccl/changefeedccl/show_changefeed_jobs_test.go#L450)):*
> ```Go
> 				t.Fatalf("Expected more rows when querying and none found for query: %s", query)
> 			}
> 		}
> ```

This block is repeated too many times; pull it into helper function?



_Suggestion:_
```Go
		if !rowResults.Next() {
			err := rowResults.Err()
			if err != nil {
				t.Fatalf("Error encountered while querying the next row: %v", err)
			} else {
				t.Fatalf("Expected more rows when querying and none found for query: %s", query)
			}
		}
		
		
```


<!-- Sent from Reviewable.io -->

sherman-grewal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76583)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, and @shermanCRL)

___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 88 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-Mw2RM1t2ewjF9cGQB3z:-Mw36vQD14QY6bZJdylR:b-igbxd7) ([raw file](https://github.com/cockroachdb/cockroach/blob/156f62553e2ff5f9e855d99aacf6f4faa948ef83/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L88)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Ooof... Arguably, I should have picked this earlier (during previous version)... But here we are.

Unfortunately creating changefeed statement node based on payload description is *unsafe* and
is broken.  Basically -- Description contains redacted string representation of the original statement,
and thus cannot be reversed (e.g. you will loose any security related information -- keys, etc). -- `changefeedJobDescription` function.

What to do:
  1. Please add a test (which I expect to fail) where you create changefeed into e.g. s3, and include some of the redactable fields there (`create changefeed ... into 's3://....?AWS_SECRET_ACCESS_KEY=my_secret'`).  Verify that
after an alter statement you still see AWS_SECRET_ACCESS_KEY=my_secret and not AWS_SECRET_ACCESS_KEY=redacted.  I expect this test to fail.  
 2. You can (and probably should) fix this issue as a separate PR on top of existing alter implementation. 
 3. The changefeed statement must be synthesized based on ChangefeedDetails fields.  Now, you don't actually have 
      to regenerate exact statement as it was typed when changefeed was created.  In particular, the thing you might have problem with is the target list.  You have target id, but the name available in ChangeFeedTarget field might no longer be available -- because the table could have been renamed.  So, you can't just assume you can put statement_names in the changefeed statement.  It would be nice if we can use tree.AliasedTableExpr in changefeed statement -- but as things are right now, that's not the case (I believe TablePattern is not the same as an alias table expr).   Perhaps the changefeed statement can be changed somehow to also allow specification of targets as table expressions? If so, you could generate table expression in the form of `[id AS name]` (this is something that select supports... you can do `select * from [30 as  namespace]` -- try running this from cockroach demo).  Other possible hacks include resolving current names for each descriptor, re-synthesizing changefeed statement using current names, and then, "renaming" them to the original statement_time_name.

</blockquote></details>

As discussed offline, there was a misunderstanding when this comment was published. In particular, @miretskiy thought that we were using the CREATE CHANGEFEED AST node (parsed by the old job description) to update the details of the existing changefeed. However, this is not the case. The AST node is only used to help generate the new job description of the updated changefeed. As a result, the SinkURI of the ChangefeedDetails remains unchanged after the alteration (leaving the un-redacted information), but sensitive information in the SinkURI remains redacted in the job description as desired.

That being said, this comment helped us catch the fact that we use the SinkURI from the AST node to perform validations, but it would be preferred to use the actual SinkURI (from the ChangefeedDetails) to perform validations instead. Thank you @miretskiy for this catch! :raised_hands: 
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 123 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-Mw2eKQz2XLINo-fLJ7Z:-Mw38SWU33l0TtAv5e-T:b9thr77) ([raw file](https://github.com/cockroachdb/cockroach/blob/156f62553e2ff5f9e855d99aacf6f4faa948ef83/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L123)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

So, now that we've fixed how  the statement is re-generated (above), I have to wonder... Would we be better off if we just take changefeed statement, and append all options there?

As far as I know, appending the same option multiple times overrides the previous occurrences.

 If you refactor changefeed statement plan hook so that  you have a separate function to compute the job record basically, then you could simply call into that function, and have it perform all of the validation, and return you the "new" job record which you will handle here.
The only thing is that function would probably need to take in a boolean argument indicating if it should update telemetry data  (yes, from changefeed, and now from here).


</blockquote></details>

See the comment above.
___
*[pkg/sql/parser/testdata/alter_changefeed, line 99 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-Mw2M7Jw9q9xIpq7J_OX:-Mw38cg75eG8kfGXEZcM:b32vhw) ([raw file](https://github.com/cockroachdb/cockroach/blob/156f62553e2ff5f9e855d99aacf6f4faa948ef83/pkg/sql/parser/testdata/alter_changefeed#L99)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

:heart: these tests. 
</blockquote></details>

:)
___
*[pkg/ccl/changefeedccl/show_changefeed_jobs_test.go, line 450 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-Mw2NP6C84ccmR2LUJSS:-Mw38gdz08-0ZJeich58:bp1olfu) ([raw file](https://github.com/cockroachdb/cockroach/blob/156f62553e2ff5f9e855d99aacf6f4faa948ef83/pkg/ccl/changefeedccl/show_changefeed_jobs_test.go#L450)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

This block is repeated too many times; pull it into helper function?
</blockquote></details>

Awesome suggestion, the code looks a lot more clean now. Thanks!! :100: 


<!-- Sent from Reviewable.io -->

miretskiy reviewed (changes requested):
> Reviewed 1 of 8 files at r1, 1 of 3 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76583)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @sherman-grewal, and @shermanCRL)

___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 123 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-Mw3ISmGDMnJtpaInqqA:-Mw3ISmGDMnJtpaInqqB:b-j534uj) ([raw file](https://github.com/cockroachdb/cockroach/blob/1271e047044b8786563658a530abcb569687bfb9/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L123)):*
> ```Go
> 		}
> 
> 		if opts.Options != nil {
> ```

Still, I would love to see if there is a way to avoid having to do this options parsing by hand so to speak;
Consider that 
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 142 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-Mw3Iva93W3FY6qLHsSz:-Mw3HyYQ2U67pjF2Z1ke:b-66v09o) ([raw file](https://github.com/cockroachdb/cockroach/blob/1271e047044b8786563658a530abcb569687bfb9/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L142)):*
> ```Go
> 					if _, ok := changefeedbase.OptionsWithNoValue[key]; ok {
> 						switch v := changefeedbase.AlterNoValueOptionType(value); v {
> 						case changefeedbase.OptUnsetValue:
> ```

So, is the expectation is that the user will specify `foo=false` if they want to unset something?
This probably needs to be comment in on (in the changefeed.Base); And probably in the PR description.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 193 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-Mw3J7QR2Fz7ktREklHy:-Mw3J7QR2Fz7ktREklHz:b-4uzzn9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1271e047044b8786563658a530abcb569687bfb9/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L193)):*
> ```Go
> 			if _, shouldProtect := details.Opts[changefeedbase.OptProtectDataFromGCOnPause]; shouldProtect && !p.ExecCfg().Codec.ForSystemTenant() {
> 				return errorutil.UnsupportedWithMultiTenancy(67271)
> 			}
> ```

I don't know; something about this code where we are re-doing work already done in create changefeed statement just doesn't sit right with me.  I do think that re-generating changefeed statement, and letting all of the validation checks happen in 1 place is a way to go.  Is there any way we can get to that point?

I honestly think the answer is ... yes.   CreateChangefeed is a simple(ish) statement:

```
type CreateChangefeed struct {
	Targets TargetList
	SinkURI Expr
	Options KVOptions
}
```

The source of truth during alter changefeed is the ChangeFeedDetails:
```
message ChangefeedDetails {
  map<uint32, ChangefeedTarget> targets
  string sink_uri = 3 [(gogoproto.customname) = "SinkURI"];
  map<string, string> opts = 4;
}
```

I believe each field in changefeed details can be converted to tree.CreateChangefeed format.
E.g. if you need to convert a stree (sink_uri) to Expr -- that's simple, tree.NewDString("....") is an expression
that evaluates to string value.  Similar for KVOptions.  Just add the new options to the list.
TargetList -- that's a bit more involved.  Basically, you don't know what the table name is right now.
But what you *do* know is the descritor ID.  So, you could resolve those descriptors to be fully qualified names right now.  Now, having created the tree.CreateChangefeed statement you should be able to call into basically the 
same code as used in changefeed_stmt.  Your goal is to get job.Record out of the function, w/out actually writing new job record.  The new job record will have correctly redacted information, etc.    The only thing you need to do after that is go over updated changefeedDetails.targets , and replace the names with the original names.

This may sound like a lot of work, but I think it's actually not.  YOu're using all of the existing validation function, you're keeping the code in 1 place; and I just think it'll make more sense.

Let me  know what you think and if you agree or not. 
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 291 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-Mw3H0Ks0ARvJVwcmBRj:-Mw3H0Ks0ARvJVwcmBRk:b-t7ik2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1271e047044b8786563658a530abcb569687bfb9/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L291)):*
> ```Go
> 		jobDescription := tree.AsString(createChangefeedStmt)
> 
> 		newPayload.Description = jobDescription
> ```

If I understand correctly, we don't need to call `changefeedJobDescription` because createChagefeedStmt was generated previous description, which should have had URI redacted.
At the very least, this deserves a comment; Better yet, I think I would prefer the entirety of taking Description, and parsing into the changefeed statement to be moved into a helper function (with comments).  something like `getUpdatedChangefeedDescription(....)`.


<!-- Sent from Reviewable.io -->

HonoreDB reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76583)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @sherman-grewal, and @shermanCRL)

___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 193 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-Mw3J7QR2Fz7ktREklHy:-Mw7liBg8MVuyU0vSAz9:b6au9m9) ([raw file](https://github.com/cockroachdb/cockroach/blob/1271e047044b8786563658a530abcb569687bfb9/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L193)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I don't know; something about this code where we are re-doing work already done in create changefeed statement just doesn't sit right with me.  I do think that re-generating changefeed statement, and letting all of the validation checks happen in 1 place is a way to go.  Is there any way we can get to that point?

I honestly think the answer is ... yes.   CreateChangefeed is a simple(ish) statement:

```
type CreateChangefeed struct {
	Targets TargetList
	SinkURI Expr
	Options KVOptions
}
```

The source of truth during alter changefeed is the ChangeFeedDetails:
```
message ChangefeedDetails {
  map<uint32, ChangefeedTarget> targets
  string sink_uri = 3 [(gogoproto.customname) = "SinkURI"];
  map<string, string> opts = 4;
}
```

I believe each field in changefeed details can be converted to tree.CreateChangefeed format.
E.g. if you need to convert a stree (sink_uri) to Expr -- that's simple, tree.NewDString("....") is an expression
that evaluates to string value.  Similar for KVOptions.  Just add the new options to the list.
TargetList -- that's a bit more involved.  Basically, you don't know what the table name is right now.
But what you *do* know is the descritor ID.  So, you could resolve those descriptors to be fully qualified names right now.  Now, having created the tree.CreateChangefeed statement you should be able to call into basically the 
same code as used in changefeed_stmt.  Your goal is to get job.Record out of the function, w/out actually writing new job record.  The new job record will have correctly redacted information, etc.    The only thing you need to do after that is go over updated changefeedDetails.targets , and replace the names with the original names.

This may sound like a lot of work, but I think it's actually not.  YOu're using all of the existing validation function, you're keeping the code in 1 place; and I just think it'll make more sense.

Let me  know what you think and if you agree or not. 
</blockquote></details>

Agree, syncing validations in multiple places seems unnecessarily painful. And hard--I think this is missing at least one, validating that you're providing a schema registry while setting the format to avro.


<!-- Sent from Reviewable.io -->

miretskiy reviewed (approved):
> Reviewed 6 of 8 files at r5, 3 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76583)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @sherman-grewal, and @shermanCRL)

___
*[-- commits, line 16 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-MwYkLll-tV06KhqyuTH):*
I think this needs to be updated to also mention UNSET
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 176 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-MwYn8K_4EaKbZnw2Cpl:-MwYn8K_4EaKbZnw2Cpm:bpurcu7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d33aa11607a62d6ba0fcc54d9dfd28946efbf23/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L176)):*
> ```Go
> 					}
> 					if _, ok := changefeedbase.AlterChangefeedUnsupportedOptions[key]; ok {
> 						return errors.Errorf(`cannot alter option %s`, key)
> ```

you could probably wrap these errors w/ e.g. 
```
return pgerror.Newf(pgcode.InvalidParameterValue, ..._ 
```

(see examples in changefeeed_stmt )
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 240 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-MwYnqIA1ztGV4rEHyup:-MwYnqIA1ztGV4rEHyuq:b3t0qrv) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d33aa11607a62d6ba0fcc54d9dfd28946efbf23/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L240)):*
> ```Go
> 		)
> 		if err != nil {
> 			return err
> ```

I think we should wrap this error to make it clear that the failure is due to "alter" .... 
errors.Wrap(err, "alter changefeed failed") or some such. 
___
*[pkg/ccl/changefeedccl/changefeedbase/options.go, line 246 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-MwYlM3L0j6qmeQK2E-R:-MwYlM3L0j6qmeQK2E-S:baoaokk) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d33aa11607a62d6ba0fcc54d9dfd28946efbf23/pkg/ccl/changefeedccl/changefeedbase/options.go#L246)):*
> ```Go
> // AlterChangefeedUnsupportedOptions are changefeed options that we do not allow
> // users to alter
> var AlterChangefeedUnsupportedOptions = makeStringSet(OptCursor, OptInitialScan, OptNoInitialScan)
> ```

Cursor?
___
*[pkg/ccl/changefeedccl/alter_changefeed_test.go, line 126 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-MwZ44iR0dLQxWJecSAs:-MwZ44iR0dLQxWJecSAt:b2dbdj3) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d33aa11607a62d6ba0fcc54d9dfd28946efbf23/pkg/ccl/changefeedccl/alter_changefeed_test.go#L126)):*
> ```Go
> 		waitForJobStatus(sqlDB, t, feed.JobID(), `paused`)
> 
> 		sqlDB.Exec(t, fmt.Sprintf(`ALTER CHANGEFEED %d SET diff`, feed.JobID()))
> ```

probably need to add unset test too.
___
*[pkg/ccl/changefeedccl/alter_changefeed_test.go, line 252 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-MwZ4Bkg0vsTpJrBGusP:-MwZ4Bkg0vsTpJrBGusQ:bqyuw2m) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d33aa11607a62d6ba0fcc54d9dfd28946efbf23/pkg/ccl/changefeedccl/alter_changefeed_test.go#L252)):*
> ```Go
> 	details, ok := job.Details().(jobspb.ChangefeedDetails)
> 	require.True(t, ok)
> ```

please add a comment on what it is we are testing (i.e. that alter changefeed does not accidently redacts secret keys)


<!-- Sent from Reviewable.io -->

miretskiy reviewed (approved):
Great job.  I have few minor nits.  Perhaps @honoredb might have more comments. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-:-MwZ4QQj4r9w8fcASnCq:b-qazutj)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @sherman-grewal, and @shermanCRL)



<!-- Sent from Reviewable.io -->

HonoreDB reviewed (approved):
> Reviewed 1 of 8 files at r1, 6 of 8 files at r5, 2 of 3 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76583)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @sherman-grewal, and @shermanCRL)



<!-- Sent from Reviewable.io -->

sherman-grewal reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/76583)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @HonoreDB, @miretskiy, @sherman-grewal, and @shermanCRL)

___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 176 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-MwYn8K_4EaKbZnw2Cpl:-MwgArh35Ae4BFIp5-zI:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d33aa11607a62d6ba0fcc54d9dfd28946efbf23/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L176)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

you could probably wrap these errors w/ e.g. 
```
return pgerror.Newf(pgcode.InvalidParameterValue, ..._ 
```

(see examples in changefeeed_stmt )
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/alter_changefeed_stmt.go, line 240 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-MwYnqIA1ztGV4rEHyup:-MwgAqutAQb1ZBrYaRAi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d33aa11607a62d6ba0fcc54d9dfd28946efbf23/pkg/ccl/changefeedccl/alter_changefeed_stmt.go#L240)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I think we should wrap this error to make it clear that the failure is due to "alter" .... 
errors.Wrap(err, "alter changefeed failed") or some such. 
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/changefeedbase/options.go, line 246 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-MwYlM3L0j6qmeQK2E-R:-MwgKpDq1pUGpHU-vODg:bx2w1g7) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d33aa11607a62d6ba0fcc54d9dfd28946efbf23/pkg/ccl/changefeedccl/changefeedbase/options.go#L246)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Cursor?
</blockquote></details>

Yeah, it's in the deny list so it will not be allowed to be altered
___
*[pkg/ccl/changefeedccl/alter_changefeed_test.go, line 126 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-MwZ44iR0dLQxWJecSAs:-MwgAXvb9dxtzik28xYu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d33aa11607a62d6ba0fcc54d9dfd28946efbf23/pkg/ccl/changefeedccl/alter_changefeed_test.go#L126)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

probably need to add unset test too.
</blockquote></details>

Done.
___
*[pkg/ccl/changefeedccl/alter_changefeed_test.go, line 252 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/76583#-MwZ4Bkg0vsTpJrBGusP:-MwgAXHD2XyPpWpobDaM:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/2d33aa11607a62d6ba0fcc54d9dfd28946efbf23/pkg/ccl/changefeedccl/alter_changefeed_test.go#L252)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

please add a comment on what it is we are testing (i.e. that alter changefeed does not accidently redacts secret keys)
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

Commit:
changefeedccl: allow users to alter changefeed options
with the ALTER CHANGEFEED statement

Currently, with the ALTER CHANGEFEED statement users
can only add or drop targets from an existing
changefeed. In this PR, we would like to extend this
functionality so that an user can edit and unset
the options of an existing changefeed as well.
The syntax of this addition is the following:

ALTER CHANGEFEED <job_id> SET <options> UNSET <opt_list>

Note that the <options> must follow the same syntax
that is used when creating a changefeed with options.
In particular, if you would like to set an option
that requires a value, you must write

SET opt = 'value'

On the other hand, if you would like to set an
option that requires no value, you must write

SET opt

Furthermore, this PR allows users to unset options.
This can be achieved by writing

UNSET <opt_list>

Where <opt_list> is a list of options that you
would like to unset. For example, if we would like
to unset the diff and resolved options for
changefeed 123, we would achieve this by writing

ALTER CHANGEFEED 123 UNSET diff, resolved

Release note (enterprise change): Added support to
the ALTER CHANGEFEED statement so that users can edit
and unset the options of an existing changefeed. The
syntax of this addition is the following:

ALTER CHANGEFEED <job_id> SET <options> UNSET <opt_list>

sherman-grewal commented:
bors r+
